docs: add AWS-supported Cilium management guide (#10256)
aws/eks-anywhere: docs/content/en/docs/clustermgmt/networking/cilium-aws-supported.md

---

> title: "AWS-Supported Cilium"

Reviewer: Marketing does not allow hyphenation with AWS. Let's say "Cilium supported by AWS"

linkTitle: "AWS-Supported Cilium"
weight: 40
description: >
  How to use and manage the AWS-supported Cilium CNI version in EKS Anywhere clusters.
---

## Overview

EKS Anywhere supports two approaches for managing Cilium CNI:

- **EKS-A Managed Cilium**: EKS Anywhere automatically manages Cilium installation and upgrades (default behavior)
- **AWS-Supported Cilium with Self-Management**: You manage Cilium deployment using Helm while using AWS-maintained images from the public ECR registry

> The AWS-supported Cilium version provides images that include security patches and bug fixes validated by Amazon, available at `public.ecr.aws/eks/cilium/cilium`.

Reviewer: ```suggestion
The AWS build of Cilium provides images that include security patches and bug fixes validated by Amazon, available at `public.ecr.aws/eks/cilium/cilium`.
```

## Transitioning to Open Source Cilium

If you want to use the full open source Cilium version with advanced features, the transition is seamless and requires no configuration changes. You can customize Cilium settings using the `helmValues` parameter (see [Helm Values Configuration for Cilium Plugin]({{< relref "../../getting-started/optional/cni#helm-values-configuration-for-cilium-plugin" >}})).

## Installing AWS-Supported Cilium with Self-Management

If you want to continue managing the CNI deployment yourself while using the AWS-supported Cilium version, you can install it using Helm.

### Prerequisites

- Helm 3.x installed
- Access to your EKS Anywhere cluster with `kubectl` configured
- Appropriate permissions to manage resources in the `kube-system` namespace

### Installation

Install the AWS-supported Cilium version using Helm (replace `1.17.6-0` with your desired version):

```bash
helm install cilium oci://public.ecr.aws/eks/cilium/cilium \
  --version 1.17.6-0 \
  --namespace kube-system \
  --set envoy.enabled=false \
  --set ingressController.enabled=false \
  --set loadBalancer.l7.backend=none
```

### Pros and Cons

**Pros:**

> - Access to full Cilium open source features with AWS support

Author: that's correct. But customer can still enable other features

- Security patches and bug fixes handled by Amazon
- Control over upgrade timing and version selection
- Greater flexibility to customize Cilium configuration

**Cons:**
- You are responsible for upgrading and validating Cilium versions
- Need to ensure compatibility between Cilium versions and Kubernetes releases

> - Additional operational overhead for monitoring and maintaining the CNI

Author: cni only supports last 3 kubernetes version. if customer doesnt upgrade their own version , they can see some issues. We are testing and validating cilium with the all kubernetes versions that we are supporting .

- No automatic upgrades during EKS Anywhere cluster upgrades

## Migrating from Open Source to AWS-Supported Version

If you're currently running the upstream open source Cilium version and want to switch to the AWS-supported version, you can perform an in-place upgrade using Helm. This migration allows you to benefit from AWS-maintained images while preserving your existing Cilium configuration and network policies.

> The AWS-supported version uses images from the AWS public ECR registry (`public.ecr.aws/eks/cilium/cilium`) and includes security patches and bug fixes validated by Amazon. This upgrade is performed as a rolling update, minimizing disruption to your workloads.

Reviewer: ```suggestion
The AWS build of Cilium uses images from the AWS public ECR registry (`public.ecr.aws/eks/cilium/cilium`) and includes security patches and bug fixes validated by Amazon. This upgrade is performed as a rolling update, minimizing disruption to your workloads.
```

> To perform the migration, run (replace `1.17.6-0` with your desired version):

Author: This will be first version in our release. But we will publish supported versions

  --version 1.17.6-0 \
  --namespace kube-system \
  --set envoy.enabled=false \
  --set ingressController.enabled=false \
  --set loadBalancer.l7.backend=none
```

After the upgrade, verify the installation status:

```bash
cilium status --wait
```

Expected output showing AWS ECR images:

```
    /¯¯\
 /¯¯\__/¯¯\    Cilium:             OK
 \__/¯¯\__/    Operator:           OK
 /¯¯\__/¯¯\    Envoy DaemonSet:    disabled (using embedded mode)
 \__/¯¯\__/    Hubble Relay:       disabled
    \__/       ClusterMesh:        disabled

DaemonSet              cilium                   Desired: 4, Ready: 4/4, Available: 4/4
Deployment             cilium-operator          Desired: 2, Ready: 2/2, Available: 2/2
Containers:            cilium                   Running: 4
                       cilium-operator          Running: 2
                       clustermesh-apiserver
                       hubble-relay
Cluster Pods:          17/17 managed by Cilium
Helm chart version:    1.17.6-0
Image versions         cilium             public.ecr.aws/eks/cilium/cilium:v1.17.6-0: 4
                       cilium-operator    public.ecr.aws/eks/cilium/operator-generic:v1.17.6-0: 2
```

> After aligning your Cilium installation with the AWS-supported version, you can optionally transition management of Cilium upgrades to EKS Anywhere. This allows EKS-A to automatically handle Cilium upgrades during cluster upgrades, reducing operational overhead.

Author: working on it .

To complete the transition:

> 1. **Upgrade Your Cluster**: Upgrade your cluster to the latest EKS Anywhere version.

Author: if they want to offload ownership , they can use first option and use aws supported cilium. 
if customer is using cilium oss , they can upgrade to aws supported cilium .
if customer is using different cni (flannel , calico , etc ), They they must remove it and install aws supported cilium

> 1. **Upgrade Your Cluster**: Upgrade your cluster to the latest EKS Anywhere version.

Author: That's correct . We are not supporting cni replacement if customer is using different CNIs

2. **Toggle the skipUpgrade Configuration**: Modify the `skipUpgrade` setting in your cluster configuration:

   ```bash
   kubectl edit cluster <cluster-name>
   ```

   Choose one of the following options:

   **Option 1**: Remove the `skipUpgrade` line entirely and use default settings:

   ```yaml
   cilium: {}
   ```

   **Option 2**: Explicitly set `skipUpgrade` to false:

   ```yaml
   cilium:
     skipUpgrade: false
   ```

3. **Save and Exit**: Save the configuration file and exit the editor.

> Once configured, EKS-A will automatically manage Cilium upgrades during cluster upgrades, ensuring compatibility and reducing manual maintenance.

Reviewer: ```suggestion
Once configured, EKS Anywhere will automatically manage Cilium upgrades during cluster upgrades, ensuring compatibility and reducing manual maintenance.
```

EKS Anywhere supports two approaches for managing the versions of the Cilium build maintained by AWS:

- **AWS build of Cilium with automatic version upgrades:**: EKS Anywhere automatically manages Cilium installation and upgrades (default behavior)
- **AWS build of Cilium with self-managed version upgrades**: You manage Cilium deployment using Helm while using AWS-maintained images from the public ECR registry

The AWS build of Cilium version provides images that include security patches and bug fixes validated by Amazon, available at `public.ecr.aws/eks/cilium/cilium`.

## Transitioning from custom CNI to AWS build of open source Cilium

If you want to use the full open source Cilium version with advanced features, the transition is seamless and requires no configuration changes. AWS provides support for the subset of Cilium features mentioned in the Overview section above. You can customize Cilium settings using the `helmValues` parameter (see [Helm Values Configuration for Cilium Plugin]({{< relref "../../getting-started/optional/cni#helm-values-configuration-for-cilium-plugin" >}})).

> ## Installing AWS build of Cilium with self-managed version upgrades

Reviewer: Let's add a section on "Installing AWS build of Cilium with automatic version upgrades". It should be brief, and simply explain that this is the default if you leave skipUpgrade set to false.

If you want to continue managing the Cilium CNI version yourself while using the AWS build of Cilium, you can install the AWS build of Cilium using Helm.

### Prerequisites

- Helm 3.x installed
- Access to your EKS Anywhere cluster with `kubectl` configured
- Appropriate permissions to manage resources in the `kube-system` namespace

### Installation

> Install the AWS build of Cilium version using Helm (replace `1.17.6-0` with your desired version):

Author: Yes we can add the ecr public address

```bash
helm install cilium oci://public.ecr.aws/eks/cilium/cilium \
  --version 1.17.6-0 \
  --namespace kube-system \
  --set envoy.enabled=false \
  --set ingressController.enabled=false \
  --set loadBalancer.l7.backend=none
```

### Pros and Cons

**Pros:**

> - Access to full Cilium open source features with AWS support

Reviewer: ```suggestion
- Access to full Cilium open source features with AWS support for a subset of features
```

- Security patches and bug fixes handled by Amazon
- Control over upgrade timing and version selection
- Greater flexibility to customize Cilium configuration

**Cons:**
- You are responsible for upgrading and validating Cilium versions

> - Need to ensure compatibility between Cilium versions and Kubernetes releases

Reviewer: This is redundant with the next bullet point. Let's remove this one.

- Additional operational overhead for validating that your Cilium version is compatible with your Kubernetes version
- No automatic upgrades during EKS Anywhere cluster upgrades

## Migrating from upstream open source Cilium to AWS build of Cilium

If you're currently running the upstream open source Cilium version and want to switch to the AWS build of Cilium, you can perform an in-place upgrade using Helm. This migration allows you to benefit from images maintained by AWS while preserving your existing Cilium configuration and network policies.

> The AWS build of Cilium version uses images from the AWS public ECR registry (`public.ecr.aws/eks/cilium/cilium`) and includes security patches and bug fixes validated by Amazon. This upgrade is performed as a rolling update, minimizing disruption to your workloads.

Reviewer: ```suggestion
The AWS build of Cilium uses images from the AWS public ECR registry (`public.ecr.aws/eks/cilium/cilium`) and includes security patches and bug fixes validated by AWS. This upgrade is performed as a rolling update, minimizing disruption to your workloads.
```
